### PR TITLE
librbd/cache/pwl/ssd/WriteLog: decrement m_bytes_allocated when retiring

### PR DESCRIPTION
**Pull request by idryomov**
Currently if ssd cache is filled to capacity, all future I/O hangs
indefinitely because even though the cache eventually becomes clean
and retires enough entries to get back under RETIRE_HIGH_WATER, this
isn't communicated to AbstractWriteLog::check_allocation().

Fixes: https://tracker.ceph.com/issues/50560
Signed-off-by: Ilya Dryomov <idryomov@gmail.com>

### CODE REVIEW

**Comment by idryomov** [comment]
cc @MahatiC

**Comment by idryomov** [comment]
cc @CongMinYin

File: src/librbd/cache/pwl/ssd/WriteLog.cc [hunk 1]
```diff
@@ -708,6 +708,8 @@ bool WriteLog<I>::retire_entries(const unsigned long int frees_per_tx) {
           m_first_valid_entry = first_valid_entry;
           ceph_assert(m_first_valid_entry % MIN_WRITE_ALLOC_SSD_SIZE == 0);
           this->m_free_log_entries += retiring_entries.size();
+          ceph_assert(this->m_bytes_allocated >= allocated_bytes);
+          this->m_bytes_allocated -= allocated_bytes;
```
Comment (CongMinYin) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:714]: This logic does exist in RWL, but it is omitted here.

Comment (idryomov) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:714]: I'm not sure I understand the comment.  Do you mean that it is omitted intentionally?

Comment (MahatiC) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:714]: No, it's not omitted intentionally. This PR looks correct. I actually noticed this issue and fixed it locally along with a few more fixes and haven't upstreamed them yet. And I am currently running some tests as well. I will review this change once again by tomorrow. Thank you for this PR!

Comment (idryomov) [reply src/librbd/cache/pwl/ssd/WriteLog.cc:714]: FWIW I have another PR for the ssd mode almost ready -- fixing power cycle issues.  Should be able to submit by tomorrow.  I'll CC you, thanks for taking a look!

**Comment by idryomov** [comment]
Added a fix for an existing free()/delete mismatch.

**Comment by MahatiC** [comment]
Thanks! Looks good to me.

**Pull request merged**